[FEAT] QnA CRUD 구현

gooot opened the pull request:
## What is this PR? 🔍

- 기능 :
* feat : QnA CRUD 구현

* feat : Question 수정 삭제
- isOwner 소유자 확인
- validModifyAndRemove 답변 유무 확인

* feat : Answer 수정 삭제
- isOnwer 소유자 확인

* feat : error 코드 추가

* feat : QuestionEntity, AnswerEntity 추가

* feat : log 추가
- issue : #25 

## Changes 📝

<!-- 이번 PR에서의 변경점 -->

## Precaution

## ✔️ Please check if the PR fulfills these requirements

- [ ] It's submitted to the correct branch, not the `develop` branch unconditionally?
- [ ] If on a hotfix branch, ensure it targets `main`?
- [ ] There are no warning message when you run `yarn lint`

- Close #ISSUE_NUMBER

gooot committed:
feat : QnA CRUD 구현

* feat : QnA CRUD 구현

* feat : Question 수정 삭제
- isOwner 소유자 확인
- validModifyAndRemove 답변 유무 확인

* feat : Answer 수정 삭제
- isOnwer 소유자 확인

* feat : error 코드 추가

* feat : QuestionEntity, AnswerEntity 추가

* feat : log 추가

limbaba1120 commented on omocha-client/src/main/java/org/auction/client/common/code/QnACode.java:
```diff
@@ -0,0 +1,49 @@
+	ANSWER_LIST_ACCESS_SUCCESS(HttpStatus.OK, "답변 목록 조회에 성공하였습니다."),
+
+	// EXPLAIN: 403 ERROR
+	QUESTION_DENY(HttpStatus.FORBIDDEN, "수정,삭제가 거부되었습니다."),
```
QUESTION_DELETE_DENY 가 더 좋을 것 같아요

ss0ngcode replied:
수정, 삭제라 이렇게 네이밍 하신거같은데
질문 답변은 수정이 안되는걸로 처리하기로 해서
수정을 없애고 건우님 네이밍으로 가는게 나을거같긴 하네요

gooot replied:
'질문의 수정 , 삭제를 불가능하게 하자' 라고 했었는데 
답변 유무에 관계없이 모두 수정 불가로 하면 될까요?

limbaba1120 replied:
수정은 불가하고 삭제는 가능한거 아닌가요?!

ss0ngcode replied:
일단은 CRUD 전부 그대로 놔두고 일단은 프론트쪽에서 사용할 api만 사용하게 하는게 좋을 것 같아요!
추후 코드를 삭제할지 말지는 논의후 정하도록 하죠

gooot replied:
우선 그대로 두었습니다

limbaba1120 commented on omocha-domain/src/main/java/org/auction/domain/qna/domain/entity/QuestionEntity.java:
```diff
@@ -0,0 +1,74 @@
+		String title,
+		String content
+	) {
+		setTitle(title);
```
setTitle() 메서드를 굳이 안만들어도 되고 바로 this.title, this.content라고 해도 될것 같아요

gooot replied:
수정했습니다

limbaba1120 commented on omocha-domain/src/main/java/org/auction/domain/qna/domain/entity/AnswerEntity.java:
```diff
@@ -0,0 +1,71 @@
+		this.content = content;
+	}
+
+	public void updateAnswer(
```
여기도 QuestionEntity에 있는 내용과 동일합니다

gooot replied:
수정했습니다

limbaba1120 commented on omocha-client/src/main/java/org/auction/client/qna/application/QuestionService.java:
```diff
@@ -0,0 +1,173 @@
+
+		isOwner(questionEntity, memberEntity);
+
+		validModifyAndRemove(questionEntity);
```
이거는 무엇을 하는 메서드인가요?

gooot replied:
답글이 달리면 '질문의 수정 , 삭제를 불가능하게 하자' 라고 했었어서 답글이 있는지 확인하는 메소드입니다.

limbaba1120 commented on omocha-client/src/main/java/org/auction/client/qna/application/AnswerService.java:
```diff
@@ -0,0 +1,162 @@
+		log.debug("remove answer finished for memberId: {} , answerId: {}", memberId, answerId);
+	}
+
+	public void isOwner(
```
is로 시작하는 메서드명 말고 조금 더 명확하게 이해할 수 있는 메서드 명을 만들면 좋을 것 같아요

ss0ngcode commented on omocha-domain/src/main/java/org/auction/domain/qna/domain/entity/AnswerEntity.java:
```diff
@@ -0,0 +1,71 @@
+	@JoinColumn(name = "question_id")
+	private QuestionEntity questionEntity;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(name = "member_id")
+	private MemberEntity memberEntity;
```
답변에 대해서는 member_id가 필요 없을 것 같아요!
해당 경매에 대한 답변은 판매자인게 확정이기 때문에 member_id는 답변 생성의 경우 경매 게시글의 seller_id와 요청한 사용자의 member_id 일치여부만 확인하는 식으로 가면 될 것 같습니다!

gooot replied:
삭제하고 questionEntity -> auctionEntity -> memberEntity -> memberId 로 우선 변경하였습니다. 
이렇게 하는게 맞는지 고민이 됩니다.

ss0ngcode commented on omocha-client/src/main/java/org/auction/client/qna/application/AnswerService.java:
```diff
@@ -0,0 +1,162 @@
+
+		log.debug("find answer started for questionId: {}", questionId);
+
+		QuestionEntity questionEntity = questionRepository.findById(questionId)
+			.orElseThrow(() -> new QnaNotFoundException(QnACode.QUESTION_NOT_FOUND));
+
+		AnswerEntity answerEntity = answerRepository.findByQuestionEntity(questionEntity)
+			.orElseThrow(() -> new QnaNotFoundException(QnACode.ANSWER_NOT_FOUND));
```
현재는 질문과 답변을 각각 따로따로 불러오는데
이것도 fetch join 같은거 사용해서 db 커넥션 두번 안거치고 한번에 불러오는 방법이 있을거 같은데...
일단은 생각이 안나니 TODO로만 남겨두고 중간발표 이후에 리팩토링하면서 찾아보도록 하죠!

gooot replied:
TODO 로 남겨두었습니다.

ss0ngcode commented on omocha-client/src/main/java/org/auction/client/qna/application/AnswerService.java:
```diff
@@ -0,0 +1,162 @@
+
+		log.debug("add answer started for memberId: {} , CreateAnswerRequest : {}", memberId, createAnswerRequest);
+
+		MemberEntity memberEntity = memberRepository.findById(memberId)
+			.orElseThrow(() -> new MemberNotFoundException(MemberCode.MEMBER_NOT_FOUND));
+
+		QuestionEntity questionEntity = questionRepository.findById(createAnswerRequest.questionId())
+			.orElseThrow(() -> new QnaNotFoundException(QnACode.QUESTION_NOT_FOUND));
+
+		AuctionEntity auctionEntity = auctionRepository.findById(questionEntity.getAuctionEntity().getAuctionId())
+			.orElseThrow(() -> new AuctionNotFoundException(AuctionCode.AUCTION_NOT_FOUND));
```
이런것도 위와 마찬가지

ss0ngcode commented on omocha-client/src/main/java/org/auction/client/qna/interfaces/QuestionController.java:
```diff
@@ -0,0 +1,142 @@
+
+	private final QuestionService questionService;
+
+	@GetMapping("/{auctionId}/question-list")
+	public ResponseEntity<ResultDto<Page<QuestionListResponse>>> questionList(
+		@PathVariable(value = "auctionId") Long auctionId,
+		@PageableDefault(page = 0, size = 10, sort = "createdAt", direction = Sort.Direction.ASC)
+		Pageable pageable
+	) {
```
질문에 대해서 list 형식으로 넘겨준다면
답변이 있는 경우를 체크해서 한번에 넘겨줄 수 있을거 같기도 하네요

예를들어 질문들을 싹 불러오고 해당 질문들에 대해서 있는지 없는지를 체크하면서 있다면 response에 질문과 짝지어서 답변을 함께 넣어주는거죠
그러면 q&a를 한 컨트롤러에서 할 수도 있을거같아요!

근데 이건 프론트분들과도 논의가 필요한 사항이라 추후 수정하도록 TODO만 남겨주세요!

gooot replied:
TODO 로 남겨두었고 어제 pageable 변경된거까지 반영했습니다.

ss0ngcode reviewed (commented):
코멘트 남겼습니다~
확인 부탁드려요!

gooot committed:
feat : PR 반영

* refactor : QuestionEntity, AnswerEntity 수정
- deleted field 추가
- update 메소드 변경
- delete 메소드 추가

* refactor : QuestionService, AnswerService 수정
- softDelete 로 변경
- 메소드명 변경

* refactor : pageable 코드 통일

* refactor : QuestionRepository, AnswerRepository
- JPA 메소드 변경

ss0ngcode commented on omocha-client/src/main/java/org/auction/client/qna/application/AnswerService.java:
```diff
@@ -0,0 +1,168 @@
+
+		log.debug("find answer started for questionId: {}", questionId);
+
+		// TODO : 추후 리팩토링 예정
+		QuestionEntity questionEntity = questionRepository.findByQuestionIdAndDeletedIsFalse(questionId)
+			.orElseThrow(() -> new QnaNotFoundException(QnACode.QUESTION_NOT_FOUND));
+
+		AnswerEntity answerEntity = answerRepository.findByQuestionEntityAndDeletedIsFalse(questionEntity)
+			.orElseThrow(() -> new QnaNotFoundException(QnACode.ANSWER_NOT_FOUND));
+
+		log.debug("find answer finished for questionId: {}", questionId);
```
이런거 TODO 남길 때 어떤 이유로 리팩토링을 할건지도 남겨주면 좋을 것 같아요!

gooot replied:
수정했습니다.

ss0ngcode commented on omocha-client/src/main/java/org/auction/client/qna/application/AnswerService.java:
```diff
@@ -0,0 +1,168 @@
+
+		log.debug("add answer started for memberId: {} , CreateAnswerRequest : {}", memberId, createAnswerRequest);
+
+		// TODO : 추후 리팩토링
+		MemberEntity memberEntity = memberRepository.findById(memberId)
+			.orElseThrow(() -> new MemberNotFoundException(MemberCode.MEMBER_NOT_FOUND));
```
여기도 위와 마찬가지

gooot replied:
추가하였습니다.

ss0ngcode commented on omocha-client/src/main/java/org/auction/client/qna/application/AnswerService.java:
```diff
@@ -0,0 +1,168 @@
+		log.debug("remove answer finished for memberId: {} , answerId: {}", memberId, answerId);
+	}
+
+	public void hasAnswerOwnership(
+		AnswerEntity answerEntity,
+		MemberEntity memberEntity
+	) {
+		if (!answerEntity.getQuestionEntity()
+			.getAuctionEntity()
+			.getMemberEntity()
+			.getMemberId()
+			.equals(memberEntity.getMemberId())) {
+			throw new InvalidMemberException(INVALID_MEMBER);
+		}
+
+	}
```
hasAuctionOwnership과 겹치는 내용이라
이 메서드를 없애고 hasAuctionOwnership에 answerEntity.getQuestionEntity().getAuctionEntity()을 넘겨주는게 더 좋아보이네요

gooot replied:
해당 메소드 삭제하고 hasAuctionOwnership 를 사용해서 확인하도록 했습니다.

limbaba1120 reviewed (approved):
확인했습니다

gooot committed:
feat : PR 반영

* chore : TODO 추가 및 수정

* refactor : AnswerService 메소드 수정
- hasAuctionOwnership 으로 통합

ss0ngcode reviewed (approved):
LGTM!
